src/Microsoft.AspNetCore.Razor.Language/RazorProjectEngine.cs

```diff
@@ -3,26 +3,63 @@
 
 using System;
 using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language.Extensions;
 
 namespace Microsoft.AspNetCore.Razor.Language
 {
     public abstract class RazorProjectEngine
     {
+        public abstract RazorConfiguration Configuration { get; }
+
         public abstract RazorProjectFileSystem FileSystem { get; }
 
         public abstract RazorEngine Engine { get; }
 
-        public abstract IReadOnlyList<IRazorProjectEngineFeature> Features { get; }
+        public IReadOnlyList<IRazorEngineFeature> EngineFeatures => Engine.Features;
+
+        public IReadOnlyList<IRazorEnginePhase> Phases => Engine.Phases;
+
+        public abstract IReadOnlyList<IRazorProjectEngineFeature> ProjectFeatures { get; }
+
+        protected abstract void ConfigureParserOptions(RazorParserOptionsBuilder builder);
+
+        protected abstract void ConfigureDesignTimeParserOptions(RazorParserOptionsBuilder builder);
 
-        public abstract RazorCodeDocument Process(RazorProjectItem projectItem);
+        protected abstract void ConfigureCodeGenerationOptions(RazorCodeGenerationOptionsBuilder builder);
 
-        public static RazorProjectEngine Create(RazorProjectFileSystem fileSystem) => Create(fileSystem, configure: null);
+        protected abstract void ConfigureDesignTimeCodeGenerationOptions(RazorCodeGenerationOptionsBuilder builder);
 
-        public static RazorProjectEngine Create(RazorProjectFileSystem fileSystem, Action<RazorProjectEngineBuilder> configure) => Create(fileSystem, RazorConfiguration.Default, configure);
+        public virtual RazorCodeDocument Process(RazorProjectItem projectItem)
+        {
+            if (projectItem == null)
+            {
+                throw new ArgumentNullException(nameof(projectItem));
+            }
+
+            return ProcessCore(projectItem, ConfigureParserOptions, ConfigureCodeGenerationOptions);
+        }
+
+        public virtual RazorCodeDocument ProcessDesignTime(RazorProjectItem projectItem)
+        {
+            if (projectItem == null)
+            {
+                throw new ArgumentNullException(nameof(projectItem));
+            }
+
+            return ProcessCore(projectItem, ConfigureDesignTimeParserOptions, ConfigureDesignTimeCodeGenerationOptions);
+        }
+
+        protected abstract RazorCodeDocument ProcessCore(
+            RazorProjectItem projectItem,
+            Action<RazorParserOptionsBuilder> configureParser,
+            Action<RazorCodeGenerationOptionsBuilder> configureCodeGeneration);
```

Comment:
This is the main spirit of this change. You configure the options - including design time - once per code document.